plugins/inputs/intel_powerstat/file.go

```diff
@@ -133,19 +135,43 @@ func (fs *fileServiceImpl) readFileToFloat64(reader io.Reader) (float64, int64,
 }
 
 // readFileAtOffsetToUint64 reads 8 bytes from passed file at given offset.
-func (fs *fileServiceImpl) readFileAtOffsetToUint64(reader io.ReaderAt, offset int64) (uint64, error) {
+func (fs *fileServiceImpl) readFileAtOffsetToUint64(reader io.ReaderAt, offset int64, timeout config.Duration) (uint64, error) {
 	buffer := make([]byte, 8)
 
 	if offset == 0 {
 		return 0, fmt.Errorf("file offset %d should not be 0", offset)
 	}
 
-	_, err := reader.ReadAt(buffer, offset)
-	if err != nil {
-		return 0, fmt.Errorf("error on reading file at offset %d: %w", offset, err)
+	if timeout == 0 {
+		_, err := reader.ReadAt(buffer, offset)
+		if err != nil {
+			return 0, fmt.Errorf("error on reading file at offset %d: %w", offset, err)
+		}
+		return binary.LittleEndian.Uint64(buffer), nil
 	}
 
-	return binary.LittleEndian.Uint64(buffer), nil
+	value := make(chan uint64)
+	errs := make(chan error)
+
+	go func() {
+		_, readErr := reader.ReadAt(buffer, offset)
+		if readErr != nil {
+			errs <- readErr
+		} else {
+			value <- binary.LittleEndian.Uint64(buffer)
+		}
+		close(value)
+		close(errs)
+	}()
+
+	select {
+	case val := <-value:
+		return val, nil
+	case err := <-errs:
+		return 0, fmt.Errorf("error on reading file at offset %d: %w", offset, err)
+	case <-time.After(time.Duration(timeout)):
+		return 0, fmt.Errorf("timeout reading file at offset %d", offset)
+	}
```

Comment:
This is a potential goroutine leak.

If `ReadAt` takes longer than given timeout, then the main goroutine will return and the created goroutine inside the function will try to send to a channel that no longer has a receiver, so it will get stuck.

Reply:
Hi @jakubsikorski, thanks for identifying it. Any idea how can I avoid the leak?

Reply:
Hey, you could use buffered channels instead, as the goroutine will no longer wait for the receiver and it will exit gracefully (GC will handle the rest).

However, closing the channels happens very soon after and using buffered channels the receiver and sender are no longer synchronized. So there might be a delay before between sender and receiver and other operations can happen in the meantime - specifically saying - closing the channel. A closed channel is always ready to communicate (with default value). This can lead to race conditions, due to non-deterministic behavior of the select statement.

Now how to solve the 2nd problem:
1. Don't close the channels. It's not necessarily a bad thing in this case, as specified in https://go.dev/tour/concurrency/4
2. Have value and error as a structure and use one channel to handle both, then choose what do it in the select based on the values in this structure.

Thanks @p-zak for finding the solution(s) :)